src/game/underwater_effects.c

```diff
@@ -0,0 +1,263 @@
+
+    u32 flags = gUnknown_03005590 & 0x7;
+    if (!flags) {
+        rand = ((u32)PseudoRandom32() & 0x300);
+        if (!rand) {
+            randX = (((u32)PseudoRandom32() & 0xF00) >> 8);
+            randY = (((u32)PseudoRandom32() & 0xF00) >> 8) - 8;
+
+            if (!(p->moveState & MOVESTATE_FACING_LEFT))
+                randX = -randX;
+
+            SpawnAirBubbles(p->x - randX, p->y - randY, p->speedAirX,
+                            ((u32)PseudoRandom32() & 0x100) >> 8);
+
+            result = TRUE;
+        }
+    }
+
+    return result;
+}
+
+static void Task_SpawnAirBubbles(void)
+{
+    TaskStrc_801F15C *ts = TaskGetStructPtr(gCurTask);
+    Sprite *s = &ts->s;
+    SpriteTransform *transform = &ts->transform;
+
+    s32 r1 = ts->x;
+    s32 r4 = ts->y;
+    s32 r2;
+    s32 unk10 = ts->unk10;
+
+    r1 += SIN((unk10 & 0xFF) * 4) >> 4;
+    {
+        struct Camera *cam = &gCamera;
+        transform->x = Q_24_8_TO_INT(r1) - cam->x;
+        transform->y = Q_24_8_TO_INT(r4) - cam->y;
+    }
+
+    r2 = ((unk10 + 1) << 4);
+    r2 = MIN(r2, 0x100);
+
+    transform->width = r2;
+    transform->height = r2;
+
+    if (ts->unk14 & 0x1)
+        transform->width = -r2;
+
+    if (ts->unk14 & 0x2)
+        transform->height = -transform->height;
+
+    if (((u16)(transform->x + 0x20) > 0x130) || ((u16)(transform->y + 0x20) > 0xE0)
+        || ((u16)gUnknown_03005660.unk0 != 1) || (gUnknown_03005660.unk4 < 0)
+        || (Q_24_8_TO_INT(r4) - 3 < gUnknown_03005660.unk4) || (ts->unk10 > 0x1E0)) {
+        TaskDestroy(gCurTask);
+        return;
+    } else {
+        ts->x += ts->unk8;
+        ts->y += ts->unkA;
+
+        ts->unk8 -= (ts->unk8 >> 3);
+        ts->unk10 += 1;
+        s->unk10 = SPRITE_FLAG(PRIORITY, 2);
+
+        s->unk10 |= (gUnknown_030054B8++ | 0x20);
+
+        sub_8004558(s);
+        sub_8004860(s, transform);
+        sub_80051E8(s);
+    }
+}
+
+static void Task_SpawnBubblesAfterDrowning(void)
+{
+    Player **refPlayer = TaskGetStructPtr(gCurTask);
+    Player *p = *refPlayer;
+
+    if (IS_ALIVE(p)) {
+        TaskDestroy(gCurTask);
+        return;
+    } else if (!(gUnknown_03005590 & (0x2 | 0x1))) {
+        if ((PseudoRandom32() & 0x300) == 0) {
+            s32 r4 = ((u32)PseudoRandom32() & 0x100) >> 8;
+
+            do {
+                // TODO: Maybe these could be converted to PseudoRandBetween?
+                u32 r3, r2, r1 = ((u32)PseudoRandom32() & 0x7FF00) >> 8;
+                r1 -= 0x400;
+
+                r2 = ((u32)PseudoRandom32() & 0x7FF00) >> 8;
+                r2 -= 0x400;
+
+                r3 = ((u32)PseudoRandom32() & 0x100) >> 8;
+
+                SpawnAirBubbles(p->x + r1, p->y + r2 - 0xC00, 0, r3);
+            } while (r4-- != 0);
+        }
+    }
+}
+
+struct Task *SpawnBubblesAfterDrowning(Player *p)
+{
+    struct Task *t
+        = TaskCreate(Task_SpawnBubblesAfterDrowning, sizeof(Player **), 0x4001, 0, NULL);
+
+    Player **refPlayer = TaskGetStructPtr(t);
```

Comment:
IMO task struct should just be

```
typedef struct {
    Player player;
}
```

The pointer pointer stuff is less likely

Reply:
I had that before.
Don't see why they wouldn't do it, but I can change it back!

Reply:
I like the idea that tasks always have a struct rather than a single value, as it's sort of like OO programming with member variables